src/controllers/engine/controllerenginejsproxy.h

```diff
@@ -0,0 +1,54 @@
+#ifndef CONTROLLERENGINEJSPROXY_H
+#define CONTROLLERENGINEJSPROXY_H
+
+#include <QJSValue>
+#include <QObject>
+
+class ControllerEngine;
+
+// An object of this class gets exposed to the JS engine, so the methods of this class
+// constitute the api that is provided to scripts under "engine" object.
+//
+// The implementation simply forwards its method calls to the ControllerEngine.
+// We cannot expose ControllerEngine directly to the JS engine because the JS engine would take
+// ownership of ControllerEngine. This is problematic when we reload a script file, because we
+// destroy the existing JS engine to create a new one. Then, since the JS engine owns ControllerEngine
+// it will try to delete it. See this Qt bug: https://bugreports.qt.io/browse/QTBUG-41171
```

Comment:
Is there a way to avoid this duplication while we wait for a fix for the QT bug? (Something like PHP's `include` directive where the text of the specified file is placed inline before compiling?) My fear is that people will forget to update these when they change ControllerEngine.

Reply:
I don't think so because every function in ControllerEngineJSProxy requires the Q_INVOKABE macro. I do not fear anyone will forget to update these because they couldn't use new functions from a script if they didn't add it to ControllerEngineJSProxy. Anyway, we're not adding anything to the API of ControllerEngine anymore because we're going to write a whole new system from scratch.